### DISCUSSION TITLE
How to delete an object in a many-to-many using postgres?

### DISCUSSION THREAD

**Discussion by mpas** [sqlalchemy/sqlalchemy#7941] [Usage Questions] [answered]
I seem to hit an issue when trying to delete an object from a many-to-many relation that seems to work with sqlite, but fails on postgres.

Many to many example:

* An Item can belong to many groups and a Group can contain many items
* I want to directly delete a group leaving the related items intact

Is it possible to delete a group containing a number of items without a need to set the Group.items collection to []

sqlite seems to work, but postgres does not..

```
sqlalchemy.exc.IntegrityError: (psycopg2.errors.ForeignKeyViolation) update or delete on table "group" violates foreign key constraint "item_group_members_group_id_fkey" on table "item_group_members"
DETAIL:  Key (id)=(1) is still referenced from table "item_group_members".```
```

```py
from sqlalchemy import Column, Integer, String, Table, ForeignKey, create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, relationship

from testcontainers.postgres import PostgresContainer


Base = declarative_base()

item_group_members = Table('item_group_members', Base.metadata,
    Column('group_id', ForeignKey('group.id'), primary_key=True),
    Column('item_id', ForeignKey('item.id'), primary_key=True)
)

class Group(Base):
    __tablename__ = 'group'

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, unique=True, nullable=False)

    # relationships
    items = relationship("Item", secondary=item_group_members, back_populates="groups")

class Item(Base):
    __tablename__ = 'item'

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, unique=False, nullable=False)

    # relationships
    groups = relationship("Group", secondary=item_group_members, back_populates="items")

def run(connection_url):
    engine = create_engine(connection_url)
    Base.metadata.create_all(engine)

    Session = sessionmaker(bind=engine)
    session = Session()

    # create some groups
    group1 = Group(name="group 1")
    group2 = Group(name="group 2")

    # create some items
    item1 = Item(name="Item 1")
    item2 = Item(name="item 2")
    item3 = Item(name="Item 1")

    # add some items to a group
    group1.items = [item1, item2, item3]
    group2.items = [item2, item3]

    # add all to the session and commit
    session.add_all([group1, group2, item1, item2, item3])
    session.commit()

    # try to delete group 1
    # FAILS when using Postgres!!!
    session.query(Group).filter_by(name="group 1").delete()

    assert session.query(Item).count() == 3
    assert session.query(Group).count() == 1

def run_sqlite():
    run('sqlite://')

def run_postgres():
    with PostgresContainer("postgres:latest") as postgres:
        run(postgres.get_connection_url())

if __name__ == '__main__':
    run_sqlite()
    run_postgres()
```

**Comment by zzzeek** [comment, marked as answer]
when you run `query(Group).delete()`, that does nothing more than emit a single "DELETE FROM group ..." statement in the database.  This does not automatically take into account rows that have a referential integrity dependency on these rows being deleted, which in this case are the rows in the `item_group_members` table.  Since your schema properly includes FOREIGN KEY constraints on `item_group_members`, PostgreSQL enforces these constraints and prohibits the deletion of rows from `group` without being told what to do about `item_group_members`.  SQLite, OTOH, does not use foreign key constraints by default, so it leaves the dangling rows inside the `item_group_members` table and from a referential integrity perspective we would compare this to basically data corruption.

Solutions:

1. use ON DELETE CASCADE for these foreign keys: 
  * https://docs.sqlalchemy.org/en/14/core/constraints.html#on-update-and-on-delete - basic background
  * https://docs.sqlalchemy.org/en/14/orm/cascades.html#using-foreign-key-on-delete-with-many-to-many-relationships - some specific notes for many-to-many relationships

2. turn on referential integrity for SQLite: https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#foreign-key-support - SQLite's ON DELETE CASCADE will do nothing if this step is not taken
3. alternatively (or in addition), use ORM delete strategies for many-to-many tables: https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#deleting-rows-from-the-many-to-many-table
4. for background on query().delete(), which is a special sub-class of "DELETE" for the ORM, see https://docs.sqlalchemy.org/en/14/orm/session_basics.html#orm-expression-update-delete , which includes notes on ON DELETE CASCADE

**Comment by mpas** [comment]
Thanks for the clear explanation!
